src/adapters/bitbucket.js

```diff
@@ -0,0 +1,212 @@
+  loadCodeTree(opts, cb) {
+    opts.path = opts.node.path
+    this._loadCodeTree(opts, (item) => {
+      if (!item.type) {
+        item.type = 'blob'
+      }
+    }, cb)
+  }
+
+  // @override
+  _getTree(path, opts, cb) {
+    this._get(`/src/${opts.repo.branch}/${path}`, opts, (err, res) => {
+      if (err) return cb(err)
+      const directories = res.directories.map((dir) => ({path: dir, type: 'tree'}))
+      res.files.forEach((file) => {
+        if (file.path.startsWith(res.path)) {
+          file.path = file.path.substring(res.path.length)
+        }
+      })
+      const tree = res.files.concat(directories)
+      cb(null, tree)
+    })
+  }
+
+  // @override
+  _getSubmodules(tree, opts, cb) {
+    if (opts.repo.submodules) {
+      return this._getSubmodulesInCurrentPath(tree, opts, cb)
+    }
+
+    const item = tree.filter((item) => /^\.gitmodules$/i.test(item.path))[0]
+    if (!item) return cb()
+
+    this._get(`/src/${opts.encodedBranch}/${item.path}`, opts, (err, res) => {
+      if (err) return cb(err)
+      // Memoize submodules so that they will be inserted into the tree later.
+      opts.repo.submodules = parseGitmodules(res.data)
+      this._getSubmodulesInCurrentPath(tree, opts, cb)
+    })
+  }
+
+  // @override
+  _getSubmodulesInCurrentPath(tree, opts, cb) {
+    const currentPath = opts.path
+    const isInCurrentPath = currentPath
+      ? (path) => path.startsWith(`${currentPath}/`)
+      : (path) => path.indexOf('/') === -1
+
+    const submodules = opts.repo.submodules
+    const submodulesInCurrentPath = {}
+    Object.keys(submodules).filter(isInCurrentPath).forEach((key) => {
+      submodulesInCurrentPath[key] = submodules[key]
+    })
+
+    // Insert submodules in current path into the tree because submodules can not
+    // be retrieved with Bitbucket API but can only by reading .gitmodules.
+    Object.keys(submodulesInCurrentPath).forEach((path) => {
+      if (currentPath) {
+        // `currentPath` is prefixed to `path`, so delete it.
+        path = path.substring(currentPath.length + 1)
+      }
+      tree.push({path: path, type: 'commit'})
+    })
+    cb(null, submodulesInCurrentPath)
+  }
+
+  // @override
+  _get(path, opts, cb) {
+    const host = location.protocol + '//' + 'api.bitbucket.org/1.0'
+    const url = `${host}/repositories/${opts.repo.username}/${opts.repo.reponame}${path || ''}`
+    const cfg  = { url, method: 'GET', cache: false }
+
+    if (opts.token) {
+      // Bitbucket App passwords can be used only for Basic Authentication.
+      // Get username of logged-in user.
+      const currentUser = JSON.parse($('body').attr('data-current-user'))
+
+      // Or get username by spliting token.
+      if (opts.token.includes(':')) {
+        const result = opts.token.split(':')
+        const username =  result[0], token = result[1]
+        cfg.headers = { Authorization: 'Basic ' + btoa(username + ':' + token) }
+      }
+      else if (currentUser.username) {
```

Comment:
Can `currentUser` be null here if the BB DOM has been changed? Maybe simplify logic:

```
let username = null
if (opts.token.includes(':')) {
   username = ...
}
else {
  const currentUser = JSON.parse($('body').attr('data-current-user'))
  if (!currentUser || !currentUser.username) {
     return cb(error_object_prompt_for_username_and_token)
  }
  username = ...
}
cfg.headers = ...
```